Create FiringGroup classes to rework how battles determine firing and targeted units

trevan opened the pull request:
This is a piece of #7823.  Currently, the firing and targeting unit detection is done in several disjointed places: `MustFightBattle#fireNavalBombardment`, `MustFightBattle#findTargetGroupsAndFire`, and `FireAa`.

This extracts the logic into functional classes (`FiringGroupFilter`) that take a `BattleState` and turn the active/alive units into a list of `FiringGroup`s.  The `FiringGroup` keeps track of which units are firing on their specific targets.

There is a `FiringGroupFilter` for bombardment, aa, and other units.  These functional classes contain the same logic you can find in `MustFightBattle#fireNavalBombardment`, `MustFightBattle#findTargetGroupsAndFire`, and `FireAa`.


<!--
  Commit comment above summarizing the update.  If multiple commits please
  summarize the change above.
  Code standards and PR guidelines can be found at:
  <https://github.com/triplea-game/triplea/wiki/Contribution-Guidelines>
-->

## Testing
<!-- Describe any manual testing performed below. -->

## Screens Shots
<!-- If there are UI updates, include screenshots below -->

## Additional Notes to Reviewer
<!-- Add any additional details that would be helpful to reviewers -->

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/wiki/PR-Release-Notes
-->

<!--RELEASE_NOTE--><!--END_RELEASE_NOTE-->

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Group of units that are firing on targets
+ *
+ * <p>If the group is suicideOnHit, then all of the units should have the same unit type
```
This seems to imply a different object semantic, perhaps even a different object. WDYT of something like `SuicideOnHitFiringGroup`? Would it simplify the eventual usages?

trevan replied:
No, it wouldn't simplify any eventual usage.  This flag is only used to determine whether to kill the units after they fire.  We have to have a separate group so that we can easily detect which units hit and so which ones can be removed.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,143 @@
+    final List<FiringGroup> groups = new ArrayList<>();
+    final List<Collection<Unit>> separatedBySuicide = separateSuicideOnHit(firingUnits);
+
+    // ensure each firing group has a unique name by adding prefixes
```
Where is the prefix added? It's an interesting comment, though scanning down it's not immediately obvious where/what this is refererring to exactly.

trevan replied:
The prefixes are added inside of the callees `generateFiringGroupsWithOneSuicideAndOneNonSuicide` and `generateFiringGroups`.

DanVanAtta replied:
Cool, though that seems to be a detail of those methods. At this level, it's not obvious that is happening, therefore this comment seems out of place. One rule of thumb, any method should be understandable by simply reading it and not having to also read the methods that it invokes. One can see if that rule is violated then there is a scaling problem to understand a given method if you have to read all of the called methods, and all of the methods that are in turn called to the n'th degree.

trevan replied:
It was there to explain why the if/else if/else.  The only reason for them is to generate suffices.  If I didn't care about having unique suffices, I would just do a simple
```
return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(name, newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

But I've been thinking about your comments here and I think if I instead change the above to:
```
return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(generateName(name, newFiringUnits, separatedBySuicide), newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

Then the logic for the unique names is more obviously contained to just dealing with the string instead of generating firing groups.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,143 @@
+      final String name,
+      final Collection<Unit> targetUnits,
+      final List<Collection<Unit>> separatedBySuicide) {
+    final List<FiringGroup> groups = new ArrayList<>();
```
WDYT of having a constructor do the work for the `isSuidideOnHit` check? 

In such a case this method would be reduced to:
```
    return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(name, newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

That makes it short enough it perhaps is worth inlining.

trevan replied:
You'll notice that the isSuicideOnHit is used to determine the suffix.  And just because it is a suicideOnHit doesn't mean it will have the same suffix.  If you look at the next method `generateFiringGroups`, it also checks the suicideOnHit but it uses a different suffix.

Here's some examples of firing unit groups and what their suffices would be:

One (doesn't matter if it is suicide or not) firing group -> No suffix
One suicide firing group and one non suicide firing group -> The suicide group has " suicide" suffix and the non suicide firing group has no suffix
Two suicide firing groups -> They will both have " suicide <unitType.name>" suffix.
Two suicide firing groups and one non suicide firing group -> The two suicide groups will have the suffix " suicide <unitType.name>" and the non suicide firing group will not have a suffix.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,143 @@
+  private static List<FiringGroup> generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+      final String name,
+      final Collection<Unit> targetUnits,
+      final List<Collection<Unit>> separatedBySuicide) {
```
This param is interesting `List<Collection<Unit>>`. I wonder a bit what the ordering of the list is. If not ordered, then should it be just a `Collection`? That makes me wonder further if this would be better as a `Map<?, Collection<Unit>>`

For map key, if an enum would perhaps do well, something like `FiringGroupType { SUICIDE_ON_HIT, STANDARD }`

WDYT?

DanVanAtta replied:
If there were a map of firing group by type, then FiringGroup would not need the `suicideOnHit` boolean.

trevan replied:
Collection is fine.  The order doesn't matter.

A map won't work because there can be more than one suicideOnHit collection.  This specific method adds a simplified prefix for the special case where there is only one suicideOnHit collection and one non suicideOnHit collection (I tried to express that in the method name).  The logic to figure out which one is the suicideOnHit and which one is not the suicideOnHit is inside of this method.  If I move that logic out of this method, then this method becomes basically nothing and would be inlined.  But doing that will increase the complexity of the caller and codeclimate will complain.

DanVanAtta replied:
>  But doing that will increase the complexity of the caller and codeclimate will complain.

codeclimate is not perfect. There are times when the actual victory is from a fundamental change rather than a surface level. In other words, sometimes right right fix to codeclimate issues are not simple, surface level fixes. If a codeclimate warning is in the way of what is actually simpler, then codeclimate should be ignored in that instance.

DanVanAtta replied:
The encoding of the 'prefix' does strike me as complicated.

What are the different types of `suicideOnHit` collection? Perhaps we are talking instead a more complicated data structure, something like: `Map<UnitType, Map<SuicideOnHitFlag, Collection<Unit>>`. Generally, such a data structure is a candidate to be converted to a first class object. One would init it by passing a collection of units, perhaps calling a factory method, and the methods on the object would be queries to return units of a given type.  

I think there could be some advantages to that:
- readily tested
- extracts a good bit of detail
- readily mocked, could be useful for testing other logic that relies on units with certain characteristics

trevan replied:
I've made changes to how this works.  Take a look at the new code.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,143 @@
+    final List<FiringGroup> groups = new ArrayList<>();
+    for (final Collection<Unit> newFiringUnits : separatedBySuicide) {
+      final boolean isSuicideOnHit =
+          Matches.unitIsSuicideOnHit().test(newFiringUnits.iterator().next());
```
Will we always be guaranteed to have a first unit in: `newFiringUnits.iterator().next()`?
I do not know, maybe there is some code earlier that guarantees this, however it seems like a pretty strong precondition that is not necessarily documented (hopefully enforced). If we convert the incoming list to a map, that would resolve this problem as you would not need to check any of the unit elements to determine if they are 'suicide on hit', you could retrieve them directly from the input map.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroupFilter.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.function.Function;
+
+/** Converts the BattleState into a list of FiringGroups */
+public interface FiringGroupFilter extends Function<BattleState, List<FiringGroup>> {}
```
Couple questions:
- Could you explain why an interface "alias" is helpful?
- Is the output list ordered?

A filter will typically exclude or include items. My presumption is that this is more of a sieve, that we are not losing any elements but instead are grouping them. Is that correct? If so, perhaps this would be better named as a `Splitter`, or as a `Sorter`.  OTOH - If we do not need an interface alias, then the naming problem would drop away.

DanVanAtta replied:
For consideration, [Effective Java 3](https://github.com/muhdkhokhar/test/blob/master/Joshua%20Bloch%20-%20Effective%20Java%20(3rd)%20-%202018.pdf), page 200:


![Screenshot from 2020-10-03 12-34-53](https://user-images.githubusercontent.com/12397753/95000358-f0d88900-0574-11eb-9695-664e1bd07bf2.png)

![Screenshot from 2020-10-03 12-36-30](https://user-images.githubusercontent.com/12397753/95000380-1bc2dd00-0575-11eb-80e6-26751a078c3f.png)

![Screenshot from 2020-10-03 12-34-35](https://user-images.githubusercontent.com/12397753/95000357-f03ff280-0574-11eb-8056-698c121fb20a.png)

DanVanAtta reviewed (commented):
There is quite a lot in this PR, perhaps about 30% reviewed

trevan commented:
I've reworked the FiringGroup.  I now use Guava's Multimap and moved the name generation to a more obvious name generation method.  I think that should resolve your feedback.

I also renamed all of the Filter classes to "Splitter" and removed the empty base interface.  And I did a few minor tweaks to `FiringGroupSplitterGeneral`: fixed some variable name and definitions, made the `canFire` variable a final variable.

trevan commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroupSplitterBombard.java:
```diff
@@ -0,0 +1,37 @@
+public class FiringGroupSplitterBombard implements Function<BattleState, List<FiringGroup>> {
+
+  @Override
+  public List<FiringGroup> apply(final BattleState battleState) {
```
The original logic that this is based off can be found at `MustFightBattle#fireNavalBombardment` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1462) and `MustFightBattle#fire` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1497).

trevan commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/aa/FiringGroupSplitterAa.java:
```diff
@@ -0,0 +1,80 @@
+  BattleState.Side side;
+
+  @Override
+  public List<FiringGroup> apply(final BattleState battleState) {
```
The original logic that this is based off of is in `MustFightBattle#updateOffensiveAaUnits` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L928), `MustFightBattle#updateDefensiveAaUnits` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L949), `FireAa#FireAa` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java#L66), and `FireAa#execute` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java#L100).

trevan commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneral.java:
```diff
@@ -0,0 +1,135 @@
+  String groupName;
+
+  @Override
+  public List<FiringGroup> apply(final BattleState battleState) {
```
The original logic that this is based off of is in `MustFightBattle#findTargetGroupsAndFire` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1638) and `MustFightBattle#fire` (https://github.com/triplea-game/triplea/blob/e160d341f6c60ecf5297aad9ab26ac5a524dc2dd/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1497)

trevan commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneral.java:
```diff
@@ -0,0 +1,135 @@
+          targetGroups.stream()
+              .filter(this.filterAirVsSubTargetGroups(enemyUnits))
+              .collect(Collectors.toList());
+      generateNamedGroups(AIR_FIRE_NON_SUBS, firingGroups, airVsSubGroups, canFire, enemyUnits);
```
This special named group will replace the "Air defend non subs" or "Air attack non subs" battle steps.  I mentioned this and posted images of what it will look like in the original PR #7823

DanVanAtta commented:
I was just opening this to review!
@trevan any plans to re-open? Breaking it up further?

trevan commented:
Oops, I was cleaning up branches and I deleted this one by mistake.

DanVanAtta commented:
All good, happily very reversible : )

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+    this.groupName = groupName;
+    this.firingUnits = firingUnits;
+    this.targetUnits = targetUnits;
+    this.suicideOnHit = this.firingUnits.stream().allMatch(Matches.unitIsSuicideOnHit());
```
It's not obvious  offhand why every unit should be 'unitIsSuicideOnHit' for the whole group to be marked as such. I'm just wondering why it's not any unit, or why it's not tracked at a per-unit level even. Perhaps a comment on the variable could explain it's role in the class.

trevan replied:
I've added comments about the suicideOnHit.  See if that works.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  public Collection<Unit> getTargetUnits() {
+    return Collections.unmodifiableCollection(targetUnits);
```
I wonder if callers could just avoid mutating the return value and/or copy as needed. If we can skip the defensive copy, we can have a performance improvement and simpler code in this class as a simpler getter would do.

trevan replied:
Something to hope for :)

DanVanAtta replied:
The copies can add up to be significant. Can we do better than hoping? The java8 streaming API is particularly helpful for operating on returned collections as the collection is not mutated.

trevan replied:
The return value of this method is passed off to other methods and those pass off to other methods.  So the code path isn't easy to follow.

But I can't remember why exactly I added unmodifiableCollection.  I thought it was because that is how the original code was but I can't find any evidence of that now.  So, I've removed the unmodifiableCollection part.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  /**
+   * Keeps alive units around
```
The concept of being kept around is not very well defined. Would you mind making this comment more specific, more defined?

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+  /**
+   * Keeps alive units around
+   *
+   * <p>Units unfortunately don't track their own status so the list of targets needs to be updated
```
I think this comment is more of an implementation note. Ideally this class is designed so that it is generic and has a role, a function, and is not aware of limitations around it. IF this class is responsible for tracking units that are still alive, perhaps that can just be cleanly documented in the class javadoc? AS-is, it's a bit odd to see this note, I'm not sure if it really answers questions, seems like if you have context then it makes more sense but otherwise is a bit confusing.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+   * @param name Name of the firing units
+   * @param firingUnits Collection of units that are firing
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
```
nit, not sure this @return javadoc adds much, restates the code. We are not publishing library code, anyone reading the javadoc will likely be reading the javadoc in the class file at the same time. I think this `@return` can perhaps be removed or otherwise should describe any invariants that would be true of the return value. For example, it is a list, what is the ordering of the list? Are there constraints on the size of the list. Also, will the firigin group be a collection of the firing units, or the target units, or both?

DanVanAtta replied:
I am a bit confused though, `suicideOnHit` already implies a specific unit type, right? If we are grouping by unit type, then don't we automatically also group by `suicideOnHit` too?

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,125 @@
+@Value
+public class FiringGroup {
+
+  private static final UnitType NON_SUICIDE_MULTIMAP_KEY = new UnitType("nonsuicide", null);
```
Creating a fake unit type seems to contradict the unit type datatype (SRP violation). The method that uses this I think probably would do better to return a first class data structure, something like:

```

class SuicideOnHitSortedFiringGroups {
     Map<UnitType, FiringGroup> suicideOnHitFiringGroups;
     Collection<Unit> nonSuicideOnHitFiringGroups;;
}
```

WDYT?

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroupSplitterBombard.java:
```diff
@@ -0,0 +1,37 @@
+
+  @Override
+  public List<FiringGroup> apply(final BattleState battleState) {
+    final Collection<Unit> canFire = battleState.getBombardingUnits();
```
nit, would 'bombarindUnits' be a more descriptive name here? Perhaps even consider inlining it to avoid the intermediate variable?

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/aa/FiringGroupSplitterAa.java:
```diff
@@ -0,0 +1,80 @@
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Creates AA and Targeted Hit firing groups
```
For the splitter class, it would be nice to see some more detail in the javadoc to explain what characteristics would be true of the output `List<FiringGroup>`. I'm not really sure what "Targeted Hit firing groups" are.

Perhaps we can/should add 'targeted hit' to the glossary?

DanVanAtta replied:
The terms are broadly not well defined. TripleA was created before the concept of Domain-Driven-Design was more mainstream, as such instead of defining and using domain terminology, brand new terminology was invented (EG: "Node").

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/aa/FiringGroupSplitterAa.java:
```diff
@@ -0,0 +1,80 @@
+            battleState.filterUnits(ALIVE, side.getOpposite()),
+            Matches.unitIsNotInfrastructure().and(Matches.unitIsBeingTransported().negate()));
+
+    final List<FiringGroup> firingGroups = new ArrayList<>();
```
The logic in here is really heavy. Do you have enough context where you could readily add some comments to guide a reader in understanding what this code is doing?

DanVanAtta replied:
Perhaps add a comment about the high level operations that are happening, what the code is roughly going to do.

trevan replied:
Check out the comments that I added to the code.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneral.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * Create firing groups with units that match {@link #firingUnitPredicate}
+ *
+ * <p>The firing groups are separated by canNotTarget, canNotBeTargetedBy, and isSuicideOnHit
```
It's a bit odd to have a generic predicate and for this class to know how the firing groups are going to be split.

Is the predicate acting as a filter to remove units from the firing groups, or is the predicate more acting at the firing group level?

The firing predicate is a bit concerning as I wonder if we are too abstract here, too complex. Could the caller apply the predicate rather than passing it in? I'm wondering if a caller would have to know a great deal about this implementation to be able to reasonably pass a good predicate to this class.

DanVanAtta replied:
Here is one thought, the list of split predicates is pretty well defined. What if we passed an enum value to specify the split type? This would I think be pretty beneficial as it would push complexity into this module, simplifying the callers. The cohesion of this class would be stronger, to understand what it's doing you would not have to look at the predicates being passed in, which is fundamentally a better and more useful module. Furthermore, by passing the right enum value, you get the right behavior, this really abstracts away implementation details from the callers. Otherwise, with a predicate a caller (or a maintainer, or someone looking to use or understand this code) has to have pretty deep knowledge of how the predicate is used.

trevan replied:
Do you mean the "filter type"?  Because the predicate is for filtering the units, it isn't for splitting the units.  But having an enum for the three different filters sounds like an ok idea.

trevan replied:
I've changed the predicate to enums that determine the predicate.

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneralTest.java:
```diff
@@ -0,0 +1,357 @@
+                    .build());
+
+    assertThat(firingGroups, hasSize(1));
+    assertThat(firingGroups.get(0).getDisplayName(), is(UNITS));
```
It seems we are testing multiple behaviors at the same time here. Is the expected display name of units simply because that is the splitter parameter? Have you considered adding a set of tests to only test display name behavior?

trevan replied:
No, the expected display name will change.  The last two tests in this file test for changed display names: `twoFiringGroupsWithCanNotTarget` and `twoGroupsWhenAirAndSeaVsSub`.

DanVanAtta replied:
I see, what I'm more pointing to is that we are testing multiple/all aspects of firingGroups per test rather than testing on behavior at a time. For example, it's considered a best practice for tests to have only one reason for why they would fail. If for example the firing group size were to change *or* if the displayName were to change, then this test would fail. At a more concrete level, isolating the behaviors makes it easier to see which parameters control which outputs.

Hence I'm suggesting to pivot the tests to group and isolate the tests that verify display name, size, etc.. and do so independently.

trevan replied:
You'll notice that the majority of the tests don't check for the display name.  Only this one and the other two I mentioned.  I understand that the test should generally only have one reason why they should fail.  And I believe this test is structured that way.

DanVanAtta replied:
Can you clarify "test should generally only have one reason why they should fail. And I believe this test is structured that way."

From an initial look, it seems like if the firing group size changes this test would fail, or if the display name changes it would fail. The size of the firing group and display name are independent of one another, which makes me think that if one of two independent things were to either change, then this test method would fail for either reason. Are the two values, the firing group size and the display name actually linked?

trevan replied:
This test is validating what a one firing group result should look like.  And a single firing group result will always have the same name.  This is a display name and the only reason to change the display name that is passed in is if there are multiple groups that need to be differentiated.

DanVanAtta reviewed (commented):
Probably about 70% of the way through. Some heavy lifting in here it seems.. Good to see the code being worked and tested :+1:

trevan commented:
I've pushed a change with more documentation.  Hopefully that helps in explaining things.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/BattleStepStrings.java:
```diff
@@ -9,11 +9,15 @@
   String CASUALTIES_SUFFIX = " casualties";
   String LAND_PARATROOPS = "Land Paratroopers";
   String NAVAL_BOMBARDMENT = "Naval bombardment";
+  String NAVAL_BOMBARD = "naval bombardment";
```
Hmm, okay. Though that seems to be overly context dependent. I wonder if it would be better to do these with built-in patterns and use them with string format. Alternatively maybe as method invocations even rather than contexts. The problem I see is that the usage is very specific, while this class is independent, it turns out to have a level of coupling to where it is used.

trevan replied:
All of these strings are only used for the Battle UI in the step list on the left hand side.  I think a probably better thing would be to use some sort of localization library and have these strings put in the language files.  I don't know if that is in any future plans.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,141 @@
+    }
+  }
+
+  String displayName;
```
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,141 @@
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
+  boolean suicideOnHit;
```
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  String displayName;
+  String groupName;
```
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,141 @@
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
```
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java:
```diff
@@ -0,0 +1,141 @@
+
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
```
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneralTest.java:
```diff
@@ -86,13 +88,60 @@ void oneWaitingToDieFiringUnitVsOneTargetableUnitMakesOneFiringGroup() {
+    final Unit fireUnit2 = givenUnitFirstStrike();
+
+    final List<FiringGroup> firingGroups =
+        FiringGroupSplitterGeneral.of(OFFENSE, FiringGroupSplitterGeneral.Type.DEFENSIVE_NORMAL, "")
```
The empty string arg seems very similar to a 'null' arg or a boolean parameter. IN general those kinds of a args are not self-explanatory and make it difficult for someone to read and use an API. Notably, you can't just use the API, you have to look up which values are allowed to then realize empty string is the way to go. One would wonder, "can I use null here"? In any case, it makes you look up the implementation which is a heuristic for a difficult to use API. Similar for those reading, they'll be unsure why an empty string is needed.

Generally the fix is to have method overloads and have an API that omits the unnecessary parameter.

DanVanAtta commented on game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneralTest.java:
```diff
@@ -210,7 +283,7 @@ void excludeSuicideOnAttackTargetsIfDefense() {
     final Unit fireUnit = givenAnyUnit();
 
     final List<FiringGroup> firingGroups =
-        FiringGroupSplitterGeneral.of(DEFENSE, unit -> true, "")
```
Looking at these usage examples, and in the prod code, it looks like the update to a flag instead of passing a predicate was a strong update, it is a much nicer looking API now :+1:
